### PR TITLE
SoundWire: fix dai prepare and hw_params management

### PR DESCRIPTION
**Pull request by plbossart**
After long hours of debug with @ujfalusi I think we finally have our root cause for the suspend/resume issues.

I have no idea how the SoundWire code might have worked in v5.10, it's one of the worst bugs I've ever seen.

@ujfalusi @bardliao @ranj063 @RanderWang thanks for your comments!

### CODE REVIEW

**Review by ranj063 — DISMISSED** [review]
Amazing find @plbossart

File: sound/soc/soc-pcm.c [hunk 1]
```diff
@@ -2023,7 +2023,7 @@ int dpcm_be_dai_hw_params(struct snd_soc_pcm_runtime *fe, int stream)
 		dev_dbg(be->dev, "ASoC: hw_params BE %s\n",
 			be->dai_link->name);
 
-		ret = __soc_pcm_hw_params(be, be_substream, &hw_params);
+		ret = __soc_pcm_hw_params(be, be_substream, &be->dpcm[stream].hw_params);
```
Comment (ujfalusi) [inline sound/soc/soc-pcm.c]: @plbossart, imho no one should save the hw_params pointer passed in hw_params() callback and afaik no one does that (with the exception of sdw intel between 5.10 and the previous patch).
From other, close to ALSA core operations it is clear that the params parameter should be only used within the hw_param() callback, it is only valid during that time.

So, I'm not sure if we want to have mixed validity of the params in driver's hw_params depending where they are in the stack.

File: drivers/soundwire/intel.c [hunk 2]
```diff
@@ -706,7 +706,6 @@ static int intel_hw_params(struct snd_pcm_substream *substream,
 	dai_runtime->paused = false;
 	dai_runtime->suspended = false;
 	dai_runtime->pdi = pdi;
-	dai_runtime->hw_params = params;
```
Comment (ujfalusi) [inline drivers/soundwire/intel.c:709]: Do we want to have a fixes tag?

Comment (plbossart) [reply drivers/soundwire/intel.c:709]: yes indeed. done

File: sound/soc/soc-pcm.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/soc-pcm.c]: I hear you, the point is not to encourage people to play with the argument. 

I thought using the heap also makes it easier to debug memory corruption issues. The stack is just awful to deal with.

Comment (ujfalusi) [reply sound/soc/soc-pcm.c]: I think we should not do this, but I will let Takashi and Mark to decide, likely going to be applied w/o much discussion ;)

My issue is that with this change a misuse will go undetected most likely and a bad example spreads faster then one would expect.

Comment (plbossart) [reply sound/soc/soc-pcm.c]: anyways I removed this patch

**Review by ujfalusi — APPROVED** [review]
Amazing detective work!!!!

**Review by ranj063 — APPROVED** [review]

**Comment by plbossart** [comment]
let's wait for @bardliao's review

**Comment by marc-hb** [comment]
Amazing war story.

For the record this was found by CI when testing these:
- thesofproject/linux/pull/4122
- thesofproject/linux/pull/4203
- thesofproject/linux/pull/4223


There are still have some suspend/resume failures in https://sof-ci.01.org/linuxpr/PR4252/build3723/devicetest/index.html and  https://sof-ci.01.org/linuxpr/PR4252/build3724/devicetest/index.html but we always had some suspend/resume failures without any audio anyway.

Also for the record: `hw_params` is allocated on the stack in `dpcm_be_dai_hw_params()`, in `sound/soc/soc-pcm.c` (thanks Pierre)

**Review by bardliao — APPROVED** [review]
Thank you @plbossart @ujfalusi for looking deep into the issue and finally find the root cause.

**Pull request merged**